hermes/probe/journal/journal.go

```diff
@@ -0,0 +1,364 @@
+
+	var reader stiface.Reader
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIGetFile, func() *probe.ProbeError {
+		r, err := fileObject.NewReader(ctx)
+		if err != nil {
+			return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readJournal(): unable to read journal file: %w", err))
+		}
+		reader = r
+		return nil
+	})
+	defer reader.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debug("readJournal(): journal file retrieved from target storage system.")
+
+	contents, readErr := ioutil.ReadAll(reader)
+	if readErr != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readJournal(): unable to read journal file: %w", readErr))
+	}
+
+	logger.Debug("readJournal(): journal file read complete.")
+	return contents, nil
+}
+
+// verifyJournal verifies that the contents of the journal file match its checksum
+// and that the journal matches the existing in-memory proto.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- nilFile: a nilFile struct with the file name and contents recorded
+//  - journal: unmarshalled journal struct, to be verified, read from the target storage system.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyJournal(ctx context.Context, file *journalFile, journal *pb.StateJournal, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	if err := file.validateChecksum(); err != nil {
+		return err
+	}
+	logger.Debug("verifyJournal(): journal  contents checksum verified against filename.")
+
+	if diff := cmp.Diff(target.Journal, journal, protocmp.Transform()); diff != "" {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): journal contents mismatch (-want +got):\n%s", diff))
+	}
+	logger.Debug("verifyJournal(): journal struct contents verified against in-memory struct.")
+	return nil
+}
+
+// resolveIntent verifies and completes the intent stored in the state journal passed.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- journal: the StateJournal read from the storage system.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func resolveIntent(ctx context.Context, journal *pb.StateJournal, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	if journal.Intent == nil {
+		return nil
+	}
+
+	fileID, err := strconv.Atoi(journal.Intent.Filename[len("Hermes_"):len("Hermes_ID")])
+	if err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, err)
```

Comment:
Annotate the error (explain what went wrong and include debugging information: for example the value of `journal.Intent.Filename`).

Reply:
I've added a more detailed annotation to the error.